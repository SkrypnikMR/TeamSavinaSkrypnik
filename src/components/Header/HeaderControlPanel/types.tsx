import React from 'react';

export interface IDiv {
    display: string;
    justifyContent: string;
    alignItems: string;
    height: string;
    width: string;
    color:string;
    marginRight: string;
    marginLeft: string;
    cursor: string;
}
