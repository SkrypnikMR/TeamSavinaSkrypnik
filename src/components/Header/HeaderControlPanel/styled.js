import styled from 'styled-components';

export const StControl = styled.div`
        display: flex;
        justify-content: flex-end;
        align-items: center;
        height: 100%;
        width: 70%;
        p{
                color:white;
        }
`;
