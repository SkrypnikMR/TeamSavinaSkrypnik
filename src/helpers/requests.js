import { routes } from '../constants/routes';

const { baseUrl } = routes;

export const getRequest = async (url, optionalHeaders = {}) => {
  const headers = Object.assign({}, optionalHeaders);
  const options = { method: 'GET', headers };
  const answer = await fetch(`${baseUrl}${url}`, options);
  if (!answer.ok) throw new Error(await answer.json());
  const result = await answer.json();
  return result;
};

export const postRequest = async (url, body, optionalHeaders = {}) => {
  const headers = Object.assign({ 'Content-Type': 'application/json' }, optionalHeaders);
  const options = {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  };
  const answer = await fetch(`${baseUrl}${url}`, options);
  return answer;
};

export const putRequest = async (url, body, optionalHeaders = {}) => {
  const headers = Object.assign({ 'Content-Type': 'application/json' }, optionalHeaders);
  const options = {
    method: 'PUT',
    headers,
    body: JSON.stringify(body),
  };
  const answer = await fetch(`${baseUrl}${url}`, options);
  const result = await answer.json();
  return result;
};
export const createAuthHeader = (support) => { return { Authorization: support.getCookie('token') }; };
