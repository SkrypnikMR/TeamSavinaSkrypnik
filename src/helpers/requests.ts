import { routes } from '../constants/routes';
import { TUrl, TBody } from './types/requestTypes';

const { baseUrl } = routes;

export const getRequest = async (url: TUrl) => {
  const options = { method: 'GET' };
  const answer = await fetch(`${baseUrl}${url}`, options);
  return answer;
};

export const postRequest = async (url : TUrl, body : TBody) : Promise<Response> => {
      const options = {
              method: 'POST',
              body: JSON.stringify(body),
              headers: { 'Content-Type': 'application/json' },
      };
  const answer = await fetch(`${baseUrl}${url}`, options);
  return answer;
};
