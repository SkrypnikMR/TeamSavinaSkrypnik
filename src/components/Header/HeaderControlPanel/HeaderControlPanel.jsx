import React from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import { StControl } from './styled';
import { HEADER_CONTROL_BTNS } from '/src/constants/componentsÐ¡onsts';
import Button from '../../UI/Button';
import { APP_ROUTES } from '/src/constants/reactRoutes';
import { ROUTS_WITHOUT_MY_ACCOUNT } from '../../../constants/ui';
import { support } from '../../../helpers/support';
import { colorDefault } from '../../UI/baseLayout';

const HeaderControlPanel = ({
    themeMode,
    setValue,
    history,
    location,
    logOut,
    userNotifSettings,
    onlineUsersCount }) => {
    const { i18n } = useTranslation();
    const handleChangeLanguage = (e) => {
        i18n.changeLanguage(e.target.value);
        localStorage.setItem('lang', e.target.value);
    };
    const handleThemeClick = ({ target }) => {
        support.setSessionStorageItem('themeMode', target.value);
        setValue({ name: 'themeMode', value: target.value });
    };
    const handleLogOutClick = () => {
        logOut();
        history.push(APP_ROUTES.login);
    };
    const handleNotifClick = (e) => {
        setValue({ name: 'settings', value: { notifications: Boolean(e.target.value) } });
        support.setSessionStorageItem('settings', { notifications: Boolean(e.target.value) });
    };
    const handleMyAccountClick = () => history.push(APP_ROUTES.account);
    const getFunctionForButtons = (el) => {
        switch (el.id) {
            case 'theme_btn': return handleThemeClick;
            case 'logOut': return handleLogOutClick;
            case 'account': return handleMyAccountClick;
            case 'notif_btn': return handleNotifClick;
            default: return handleChangeLanguage;
        }
    };
    return (
        <StControl >
            <p>
                {location.pathname === '/chat'
                    ? `${i18n.t('online')}: ${onlineUsersCount === 0 ? 0 : onlineUsersCount - 1}`
                    : null}
            </p>
            {' '}
            {HEADER_CONTROL_BTNS.map((el) => {
                if (el.value === themeMode) return null;
                if (el.id === 'notif_btn' && Boolean(el.value) === userNotifSettings) return null;
                if ((el.rout === '/account' && ROUTS_WITHOUT_MY_ACCOUNT.includes(location.pathname))
                    || el.rout === location.pathname) return null;
                return (
                    <Button
                        id={el.id}
                        content={el.content}
                        key={el.id}
                        color={colorDefault}
                        fontSize='26px'
                        width='60px'
                        height="10vh"
                        borderRadius="0px"
                        value={el.value}
                        bgColor="transparent"
                        onClick={getFunctionForButtons(el)}
                    />
                );
            })}
        </StControl>
    );
};

HeaderControlPanel.propTypes = {
    setValue: PropTypes.func,
    themeMode: PropTypes.string,
    history: PropTypes.object,
    location: PropTypes.object,
    logOut: PropTypes.func,
    userNotifSettings: PropTypes.bool,
    onlineUsersCount: PropTypes.number,
};

export default HeaderControlPanel;
