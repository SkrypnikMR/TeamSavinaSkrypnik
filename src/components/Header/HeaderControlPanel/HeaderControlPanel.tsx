import React from 'react';
import { useTranslation } from 'react-i18next';
import { APP_ROUTES } from 'src/constants/reactRoutes';

import { HEADER_CONTROL_BTNS } from 'src/constants/componentsÐ¡onsts';
import Button from '../../UI/Button';
import { support } from '../../../helpers/support';
import { colorDefault } from '../../UI/baseLayout';
import { useTheme } from 'src/components/Hook/useTheme';

import { StControl } from './styled';

const HeaderControlPanel = ({
    history,
    location }) => {
    const { i18n } = useTranslation();
    const { changeTheme } = useTheme();
    const handleChangeLanguage = (e) => {
        i18n.changeLanguage(e.target.value);
        localStorage.setItem('lang', e.target.value);
    };
    const handleThemeClick = ({ target }) => {
        changeTheme();
        // support.setSessionStorageItem('themeMode', target.value);
    };
    const handleLogOutClick = () => {
        history.push(APP_ROUTES.login);
    };
    const handleNotifClick = (e) => {
        support.setSessionStorageItem('settings', { notifications: Boolean(e.target.value) });
    };
    const handleMyAccountClick = () => history.push(APP_ROUTES.account);
    const getFunctionForButtons = (el) => {
        switch (el.id) {
            case 'theme_btn': return handleThemeClick;
            case 'logOut': return handleLogOutClick;
            case 'account': return handleMyAccountClick;
            case 'notif_btn': return handleNotifClick;
            default: return handleChangeLanguage;
        }
    };
    return (
        <StControl >
            {' '}
            {HEADER_CONTROL_BTNS.map((el) => {
                return (
                    <Button
                        id={el.id}
                        content={el.content}
                        key={el.id}
                        color={colorDefault}
                        fontSize='26px'
                        width='40px'
                        borderRadius="0px"
                        value={el.value}
                        bgColor="transparent"
                        onClick={getFunctionForButtons(el)}
                    />
                );
            })}
            <img src="../../../../public/assets/images/log-out.png"/>
        </StControl>
    );
};

export default HeaderControlPanel;
