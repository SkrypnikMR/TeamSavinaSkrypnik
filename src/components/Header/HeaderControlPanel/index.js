import { withRouter } from 'react-router';
import HeaderControlPanel from './HeaderControlPanel.jsx';
import { connect } from 'react-redux';
import { setValue } from '/src/Store/user/actions';
import { userThemeMode, userNotifSettings } from '/src/Store/user/selectors';
import { logOutAction } from '/src/Store/chat/actions';
import { onlineUsersCount } from '/src/Store/chat/selectors';

const mapStateToProps = state => ({
    onlineUsersCount: onlineUsersCount(state),
    themeMode: userThemeMode(state),
    userNotifSettings: userNotifSettings(state),
});
const mapDispatchToProps = dispatch => ({
    setValue: payload => dispatch(setValue(payload)),
    logOut: () => dispatch(logOutAction()),
});

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(HeaderControlPanel));
