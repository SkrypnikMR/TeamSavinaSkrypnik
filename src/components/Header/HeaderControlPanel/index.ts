import { withRouter } from 'react-router';
import HeaderControlPanel from './HeaderControlPanel';

export default withRouter(HeaderControlPanel);
