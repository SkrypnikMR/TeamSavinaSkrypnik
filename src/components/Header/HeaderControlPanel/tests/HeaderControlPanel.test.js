import React from 'react';
import { shallowSmart, mountSmart } from '/src/helpers/testHelper';
import { useTranslation } from 'react-i18next';
import HeaderControlPanel from '../HeaderControlPanel';

jest.mock('react-i18next', () => ({
    useTranslation: jest.fn()
        .mockReturnValue({ i18n: { changeLanguage: jest.fn() }, t: jest.fn() }),
}));

describe('HeaderControlPanel', () => {
    let props;
    beforeEach(() => {
        props = {
            setValue: jest.fn(),
            themeMode: '',
            history: [],
            location: { pathname: '/' },
      };
    });
    const { i18n } = useTranslation();
    it('Should match snapshot', () => {         
        const component = shallowSmart(<HeaderControlPanel {...props}/>);
        expect(component.html()).toMatchSnapshot();
    });
    it('should call i18n.changeLanguage ', () => {
        const component = mountSmart(<HeaderControlPanel {...props}/>);
        component.find('Button').at(2).getElement().props.onClick({ target: { value: 'light' } });
        expect(i18n.changeLanguage).toHaveBeenCalled();
    });
    it('should call button for change theme, themeMode===dark', () => {
        props = {
            setValue: jest.fn(),
            themeMode: 'dark',
            history: [],
            location: { pathname: '/' },
        };
        const component = mountSmart(<HeaderControlPanel {...props}/>);
        component.find('Button').at(0).simulate('click', { target: { value: 'light' } });
        expect(props.setValue).toHaveBeenCalledWith({ name: 'themeMode', value: 'light' });
    });
    it('should call button for change theme, themeMode===light', () => {
        props = {
            setValue: jest.fn(),
            themeMode: 'light',
            history: [],
            location: { pathname: '/' },
        };
        const component = mountSmart(<HeaderControlPanel {...props}/>);
        component.find('Button').at(0).simulate('click', { target: { value: 'dark' } });
        expect(props.setValue).toHaveBeenCalledWith({ name: 'themeMode', value: 'dark' });
    });
    it('should render all buttons in /chat', () => {
        const props = {
            history: [],
            location: { pathname: '/chat' },
            setValue: jest.fn(),
            themeMode: 'light',
        };
        const component = mountSmart(<HeaderControlPanel {...props} />);
        expect(component.find('button')).toHaveLength(5);
    });
    it('should render all buttons in /account', () => {
        const props = {
            history: [],
            location: { pathname: '/account' },
            setValue: jest.fn(),
            themeMode: 'light',
        };
        const component = mountSmart(<HeaderControlPanel {...props} />);
        expect(component.find('button')).toHaveLength(4);
    });
    it('should call history push', () => {
        const props = {
            history: { push: jest.fn() },
            location: { pathname: '/chat' },
            setValue: jest.fn(),
            themeMode: 'light',
        };
        const component = mountSmart(<HeaderControlPanel {...props} />);
        component.find('Button').at(4).getElement().props.onClick();
        expect(props.history.push).toHaveBeenCalledWith('/account');
    });
});
