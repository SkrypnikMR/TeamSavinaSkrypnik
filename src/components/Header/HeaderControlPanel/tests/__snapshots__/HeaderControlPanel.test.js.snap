// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderControlPanel Should match snapshot 1`] = `"<div class=\\"styled__StControl-sc-1i99xqk-0 iDtmsZ\\"><p></p> <button id=\\"notif_btn\\" type=\\"button\\" value=\\"1\\" color=\\"white\\" font-size=\\"26px\\" height=\\"10vh\\" width=\\"60px\\" class=\\"styled__StButton-sc-1a6hck8-0 jFhuDa\\"></button><button id=\\"theme_btn\\" type=\\"button\\" value=\\"light\\" color=\\"white\\" font-size=\\"26px\\" height=\\"10vh\\" width=\\"60px\\" class=\\"styled__StButton-sc-1a6hck8-0 jFhuDa\\"></button><button id=\\"theme_btn\\" type=\\"button\\" value=\\"dark\\" color=\\"white\\" font-size=\\"26px\\" height=\\"10vh\\" width=\\"60px\\" class=\\"styled__StButton-sc-1a6hck8-0 jFhuDa\\"></button><button id=\\"en\\" type=\\"button\\" value=\\"en\\" color=\\"white\\" font-size=\\"26px\\" height=\\"10vh\\" width=\\"60px\\" class=\\"styled__StButton-sc-1a6hck8-0 jFhuDa\\"></button><button id=\\"ru\\" type=\\"button\\" value=\\"ru\\" color=\\"white\\" font-size=\\"26px\\" height=\\"10vh\\" width=\\"60px\\" class=\\"styled__StButton-sc-1a6hck8-0 jFhuDa\\"></button><button id=\\"ar\\" type=\\"button\\" value=\\"ar\\" color=\\"white\\" font-size=\\"26px\\" height=\\"10vh\\" width=\\"60px\\" class=\\"styled__StButton-sc-1a6hck8-0 jFhuDa\\"></button></div>"`;
